**Model building using Decision Tree Classifier**

In [1]:
import pandas as pd

In [2]:
from google.colab import files
uploaded = files.upload()

Saving diabetes.csv to diabetes.csv


In [4]:
# Reading the dataset from the csv file
# separator is a vertical line, as seen in the dataset
data = pd.read_csv("diabetes.csv")

# Printing the shape of the dataset
print(data.shape)

(768, 9)


In [5]:
data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [30]:
X = data.drop("Outcome",axis="columns")
y = data.Outcome

In [31]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [32]:
X_scaled[:3]

array([[ 0.63994726,  0.84832379,  0.14964075,  0.90726993, -0.69289057,
         0.20401277,  0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575,  0.53090156, -0.69289057,
        -0.68442195, -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, -1.28821221, -0.69289057,
        -1.10325546,  0.60439732, -0.10558415]])

In [33]:
from sklearn.model_selection import train_test_split

In [34]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, stratify=y, random_state=10)

In [29]:
X_train.shape

(576, 9)

In [15]:
X_test.shape

(192, 8)

In [16]:
from sklearn.tree import DecisionTreeClassifier

In [17]:
from sklearn.model_selection import cross_val_score

In [18]:
scores = cross_val_score(DecisionTreeClassifier(), X, y, cv=5)
scores

array([0.7012987 , 0.66233766, 0.69480519, 0.77777778, 0.7124183 ])

In [19]:
scores.mean()

0.7097275273745862

Using the cross-validation score, we get the accuracy score to be 0.7097 (Approximately). We can build the same model using the bagging algorithms to compare the accuracy scores.

**Implementing the Bagging algorithms**

In [20]:
from sklearn.ensemble import BaggingClassifier

In [21]:
bag_model = BaggingClassifier(
base_estimator=DecisionTreeClassifier(), 
n_estimators=100, 
max_samples=0.8, 
bootstrap=True,
oob_score=True,
random_state=0
)

In [22]:
bag_model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


BaggingClassifier(base_estimator=DecisionTreeClassifier(), max_samples=0.8,
                  n_estimators=100, oob_score=True, random_state=0)

In [23]:
bag_model.oob_score_

0.7534722222222222

The model improves the accuracy score. 

We can also check the accuracy score using the testing dataset to determine if our model is overfitting. To get the accuracy score, use this code:

In [24]:
bag_model.score(X_test, y_test)

0.7760416666666666

The accuracy score shows that our model is not overfitting. Overfitting occurs when we get a lower accuracy when using the testing dataset.